### PR TITLE
C++14 port

### PR DESCRIPTION
**Pull request by jbytheway**
Porting to C++14 because the dependent libraries no longer work in C++03 on my system.  Entailed a variety of small changes, due to e.g. nuances in what gets included by which headers, new compile flags, and syntax changes in the new C++.

### CODE REVIEW

**Comment by jbytheway** [comment]
Well, it took a lot of tweaking to fix all the issues from the various slight discrepancies between my build environment and Travis's, but I think it's fixed now.  I would be curious to know if I broke things for other environments before I merge.

**Comment by m50d** [comment]
`make clean` is failing for me which is not a great start. Will try to pursue by the end of the week.

**Comment by jbytheway** [comment]
Alas.  Feel free to send the output my way if it's not an obvious fix.

**Comment by m50d** [comment]
Looks like it wasn't deleting some broken symlinks - I suspect an issue with my platform rather than anything on the sakusen side.

File: .travis.yml [hunk 1]
```diff
@@ -4,10 +4,14 @@ addons:
   apt:
     sources:
       - deadsnakes
+      - ubuntu-toolchain-r-test
+      - sourceline: 'deb http://archive.ubuntu.com/ubuntu trusty-backports main restricted universe multiverse'
```
Comment (m50d) [inline .travis.yml]: I take it we need something that isn't available in the structured "sources"

**Comment by m50d** [comment]
Most of the build was fine but I got errors on the TCL bindings that suggest they're not getting built with the correct `-std` or some such:

````
make[3]: Entering directory '/home/michaeld/sakusen/bindings/tcl/libsakusen'
swig -DSWIGWORDSIZE64 -tcl -c++ -o sakusen.cpp -I../../../ -I/usr/include/tcl ../../../bindings/libsakusen.i
../../../sakusen/point.h:172: Warning 378: operator!= ignored
../../../sakusen/point.h:219: Warning 365: operator+= ignored
../../../sakusen/point.h:232: Warning 366: operator-= ignored
../../../sakusen/point.h:243: Warning 367: operator*= ignored
../../../sakusen/point.h:248: Warning 368: operator/= ignored
../../../sakusen/ref.h:112: Warning 302: Identifier 'operator ==' redefined (ignored),
../../../sakusen/ref.h:21: Warning 302: previous definition of 'operator =='.
../../../sakusen/box.h:89: Warning 365: operator+= ignored
../../../sakusen/idbase.h:97: Warning 378: operator!= ignored
../../../sakusen/perception.h:25: Warning 372: operator|= ignored
../../../sakusen/hitpoints.h:51: Warning 365: operator+= ignored
../../../sakusen/hitpoints.h:68: Warning 366: operator-= ignored
../../../sakusen/hitpoints.h:82: Warning 367: operator*= ignored
../../../sakusen/hitpoints.h:99: Warning 367: operator*= ignored
../../../sakusen/hitpoints.h:113: Warning 368: operator/= ignored
../../../sakusen/hitpoints.h:127: Warning 368: operator/= ignored
../../../sakusen/orientation.h:52: Warning 378: operator!= ignored
../../../sakusen/frame.h:36: Warning 365: operator+= ignored
../../../sakusen/point.h:149: Warning 503: Can't wrap 'operator []' unless renamed to a valid identifier.
../../../sakusen/point.h:149: Warning 503: Can't wrap 'operator []' unless renamed to a valid identifier.
../../../sakusen/point.h:149: Warning 503: Can't wrap 'operator []' unless renamed to a valid identifier.
../../../sakusen/point.h:149: Warning 503: Can't wrap 'operator []' unless renamed to a valid identifier.
../../../sakusen/point.h:149: Warning 503: Can't wrap 'operator []' unless renamed to a valid identifier.
../../../sakusen/idbase.h:47: Warning 503: Can't wrap 'operator unsigned char' unless renamed to a valid identifier.
../../../sakusen/idbase.h:92: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:95: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:96: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/idbase.h:47: Warning 503: Can't wrap 'operator unsigned char' unless renamed to a valid identifier.
../../../sakusen/idbase.h:92: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:95: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:96: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/idbase.h:47: Warning 503: Can't wrap 'operator unsigned int' unless renamed to a valid identifier.
../../../sakusen/idbase.h:92: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:95: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:96: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/idbase.h:47: Warning 503: Can't wrap 'operator unsigned int' unless renamed to a valid identifier.
../../../sakusen/idbase.h:92: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:95: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:96: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/idbase.h:47: Warning 503: Can't wrap 'operator unsigned int' unless renamed to a valid identifier.
../../../sakusen/idbase.h:92: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:95: Warning 503: Can't wrap 'operator ++' unless renamed to a valid identifier.
../../../sakusen/idbase.h:96: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/rectangle.h:126: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/rectangle.h:126: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/hitpoints.h:129: Warning 503: Can't wrap 'operator <<' unless renamed to a valid identifier.
../../../sakusen/hitpoints.h:130: Warning 503: Can't wrap 'operator >>' unless renamed to a valid identifier.
../../../sakusen/ref.h:43: Warning 503: Can't wrap 'operator MPtr' unless renamed to a valid identifier.
../../../sakusen/ref.h:51: Warning 503: Can't wrap 'operator bool' unless renamed to a valid identifier.
../../../sakusen/ref.h:52: Warning 503: Can't wrap 'operator !' unless renamed to a valid identifier.
../../../sakusen/ref.h:43: Warning 503: Can't wrap 'operator MPtr' unless renamed to a valid identifier.
../../../sakusen/ref.h:51: Warning 503: Can't wrap 'operator bool' unless renamed to a valid identifier.
../../../sakusen/ref.h:52: Warning 503: Can't wrap 'operator !' unless renamed to a valid identifier.
../../../sakusen/ref.h:43: Warning 503: Can't wrap 'operator MPtr' unless renamed to a valid identifier.
../../../sakusen/ref.h:51: Warning 503: Can't wrap 'operator bool' unless renamed to a valid identifier.
../../../sakusen/ref.h:52: Warning 503: Can't wrap 'operator !' unless renamed to a valid identifier.
../../../sakusen/sensorreturns.h:112: Warning 503: Can't wrap 'operator ()' unless renamed to a valid identifier.
../../../sakusen/box.h:54: Warning 503: Can't wrap 'operator []' unless renamed to a valid identifier.
../../../sakusen/box.h:80: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/box.h:97: Warning 503: Can't wrap 'operator <=' unless renamed to a valid identifier.
../../../sakusen/orientation.h:47: Warning 503: Can't wrap 'operator ()' unless renamed to a valid identifier.
../../../sakusen/orientation.h:49: Warning 503: Can't wrap 'operator ==' unless renamed to a valid identifier.
../../../sakusen/ordertype.h:42: Warning 509: Overloaded method sakusen::operator <<(std::ostream &,sakusen::OrderType) effectively ignored,
../../../sakusen/resourcesearchresult.h:26: Warning 509: as it is shadowed by sakusen::operator <<(std::ostream &,sakusen::ResourceSearchResult).
../../../sakusen/updatetype.h:61: Warning 509: Overloaded method sakusen::operator <<(std::ostream &,sakusen::UpdateType) effectively ignored,
../../../sakusen/resourcesearchresult.h:26: Warning 509: as it is shadowed by sakusen::operator <<(std::ostream &,sakusen::ResourceSearchResult).
../../../sakusen/resourcesearchresult.h:26: Warning 509: Overloaded method sakusen::operator <<(std::ostream &,sakusen::ResourceSearchResult) effectively ignored,
../../../sakusen/ordertype.h:42: Warning 509: as it is shadowed by sakusen::operator <<(std::ostream &,sakusen::OrderType).
../../../sakusen/world.h:71: Warning 454: Setting a pointer/reference variable may leak memory.
libtool --mode=compile --tag=CXX g++  -I../../../ -I/usr/include/tcl \
        -c -o sakusen.o sakusen.cpp
libtool: compile:  g++ -I../../../ -I/usr/include/tcl -c sakusen.cpp  -fPIC -DPIC -o .libs/sakusen.o
In file included from /usr/include/c++/5/type_traits:35:0,
                 from ../../../sakusen/point.h:5,
                 from sakusen.cpp:2199:
/usr/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
 #error This file requires compiler and library support \
  ^
````

File: .travis.yml (hunk 1 above)
Comment (jbytheway) [inline .travis.yml]: My Travis-fu is weak, but I couldn't find trusty-backports in [the whitelist](https://github.com/travis-ci/apt-source-whitelist/blob/master/ubuntu.json) (I guess that's what you're referring to?).  There may be another way to get a swig 3, but this seemed like a reasonable one.

**Comment by jbytheway** [comment]
Indeed the TCL bindings needed that.  I didn't catch it because I wasn't building them.  Travis didn't catch it because C++14 is the default on gcc-6.  I suppose we ought to make Travis build with gcc versions 5 and 6 (and clang?) to catch that sort of thing.

**Comment by m50d** [comment]
Clean build/test for me with those updates, happy for this to be merged

**Pull request merged**